docs/guides/relay.md

```diff
@@ -0,0 +1,381 @@
+---
+title: Relay
+---
+
+# Relay Guide
+
+## What is Relay?
+
+The relay spec defines some interfaces that GraphQL servers can follow to allow
+clients to interact with them in a more efficient way. The spec makes two
+core assumptions about a GraphQL server:
+
+1. It provides a mechanism for refetching an object
+2. It provides a description of how to page through connections.
+
+You can read more about the relay spec
+[here](https://relay.dev/docs/en/graphql-server-specification/).
+
+### Relay implementation example
+
+Suppose we have the following type:
+
+```python
+@strawberry.type
+class Fruit:
+    name: str
+    weight: str
+```
+
+We want it to have a globally unique ID, a way to retrieve a paginated results
+list of it and a way to refetch if if necessary. For that, we need to inherit it
+from the `Node` interface and implement its abstract methods: `resolve_id`,
+`resolve_node` and `resolve_nodes`.
+
+```python
+@strawberry.type
+class Fruit(strawberry.relay.Node):
+    code: strawberry.Private[int]
+    name: str
+    weight: float
+
+    @classmethod
+    def resolve_id(cls, root: Self, *, info: Optional[Info] = None):
+        return root.code
+
+    @classmethod
+    def resolve_nodes(
+        cls,
+        *,
+        info: Optional[Info] = None,
+        node_ids: Optional[Iterable[str]] = None,
+    ):
+        if node_ids is not None:
+            return [fruits[nid] for nid in node_ids]
+
+        return list(fruits.values())
+
+    @classmethod
+    def resolve_node(
+        cls,
+        node_id: str,
+        *,
+        info: Optional[Info] = None,
+        required: bool = False,
+    ):
```

Comment:
are these methods always required?

Reply:
Yes, basically. I even defined then as `abstractmethods` in the `Node` interface.

The idea here is that any type implementing `Node` should define 3 methods:

1) `resolve_id`, which should return the real id of the object to be used to generate the global id. I would love to allow the type to use an `id` directly, but then there's no way to have it use the resolver (at least not without some refactoring to allow that)

2) `resolve_nodes`, which is expected to return an iterable of available nodes so that the connection can paginate over it

3) `resolve_node`, which should return the node given its id. During the migration I though about removing this one, since we can basically call `resolve_nodes` and pass `node_ids=[node_id]`. What do you think?

Reply:
Also note that integrations can define those automatically.

For example, in strawberry-django-plus, when a type is created and it implements `Node`, [I inject those resolvers into the type with async-safe resolvers](https://github.com/blb-ventures/strawberry-django-plus/blob/main/strawberry_django_plus/type.py#L293) that queries the django model, and also applying the query optimizer, etc. So anyone using the django integration there can just inherit from `Node` and not worry about those methods (unless he wants to override them for some reason)

Reply:
> resolve_id, which should return the real id of the object to be used to generate the global id. I would love to allow the type to use an id directly, but then there's no way to have it use the resolver (at least not without some refactoring to allow that)

I think we should find a way to make this work 😊

For `resolve_nodes` and `resolve_node`, I think it would be nice to have only one, but at the same time I wonder if `resolve_nodes` is needed all the time, for example if you do something like this:

```python
@strawberry.type
class Website:
    id: strawberry.ID
    domain: str

    @classmethod
    def from_model(cls, model: models.Website) -> Self:
        return cls(id=model.pk, domain=model.domain)

    @strawberry.relay.connection
    async def feedbacks(
        self, info: Info[Context, None]
    ) -> strawberry.relay.Connection[Feedback]:
        feedbacks = await find_feedbacks_for_site(self.domain)

        return strawberry.relay.Connection(
            edges=[
                strawberry.relay.Edge.from_node(Feedback.from_model(feedback))
                for feedback in feedbacks
            ],
            page_info=strawberry.relay.PageInfo(
                has_next_page=False,
                has_previous_page=False,
                start_cursor="",
                end_cursor="",
            ),
        )
```

I'm already passing the nodes to the connection, right? happy to chat about this on discord too :)

Reply:
> I think we should find a way to make this work :blush:

![download (3)](https://user-images.githubusercontent.com/134025/218112372-48e0b706-b849-4c62-b997-3b67cbbfa075.png)

> I'm already passing the nodes to the connection, right?

In that specific case I would argue that you could use `@strawberry.field` directly, because you are not using any of the "magic" behind `@strawberry.relay.connection`

The main use cases for the connection field are those:

1) Define the field as `feedbacks: Connection[Feedback]`. The field would be created with the `first`, `last`, `before` and `after` arguments, call `resolve_nodes` and use those arguments to properly paginate it.

2) Use `@strawberry.relay.connection` to return an `Iterable[Feedback]`, which is basically your way of giving the paginator your own nodes so that it doesn't have to call `resolve_nodes`. In the resolver you could even define some extra args for filtering/ordering the results, but forget about the pagination stuff (which is the `Connection`'s responsibility). Just like this example in the tests: https://github.com/strawberry-graphql/strawberry/pull/2511/files#diff-089578b7231144ec508d0023bb9562b896c8ef0e24ed10eabadfe7409028886dR161

A real example for this that I use: I have a django type that can be soft deleted. Since `resolve_nodes` would return the queryset with all objects, even those that were "deleted", i pre-filter the deleted ones in the resolver. For example:

```python
@strawberry.relay.connection
def fruits(self, include_removed: bool = False) -> Iterable[FruitType]:
    return Fruit.objects.filter(is_removed=not include_removed).order_by("name")
```

This would still generate this in the schema:

```graphql
type Query {
  fruits (
    first: Int
    last: Int
    before: String
    after: String
    includeRemoved: Bool! = false
  ): FruitTypeConnection!
}
```

So paginating over that connection would retrieve only results that are not removed (unless passing `includeRemoved: true`) and ordered by name

But having said all of that, I can totally not make them "mandatory", specially because the user can always use just the resolver approach, meaning `resolve_nodes` would not be used in this case. I can probably check when creating the schema that, when using the approach number 1 the `Node` implemented type has a `resolve_nodes` and give a warning to the user. What do you think?

> happy to chat about this on discord too :)

Sure! Always open to chat about anything on discord :)

Reply:
All resolved